src/frontend/src/optimizer/plan_node/logical_filter.rs

```diff
@@ -185,11 +190,129 @@ impl ToBatch for LogicalFilter {
     }
 }
 
+/// Apply filters by selectivity and then applicabiliy of watermark - equality condition
+/// first, then conditions of the form T > now() - Y (the timestamp needs to be greater
+/// than a watermark), then conditions similar to T < now() - Y
+fn convert_comparator_to_priority(comparator: Type) -> i32 {
+    match comparator {
+        Type::Equal => 0,
+        Type::GreaterThan | Type::GreaterThanOrEqual => 1,
+        Type::LessThan | Type::LessThanOrEqual => 2,
+        _ => -1,
+    }
+}
+
 impl ToStream for LogicalFilter {
     fn to_stream(&self) -> Result<PlanRef> {
         let new_input = self.input().to_stream()?;
-        let new_logical = self.clone_with_input(new_input);
-        Ok(StreamFilter::new(new_logical).into())
+
+        let predicate = self.predicate();
+        let has_now = predicate
+            .conjunctions
+            .iter()
+            .any(|cond| cond.count_nows() > 0);
+        if has_now {
+            let mut conjunctions = predicate.conjunctions.clone();
+            // Check if the now expr is valid
+            for conjunction in &conjunctions {
+                if conjunction.count_nows() > 0 {
+                    let comparator_expr = try_match_expand!(conjunction, ExprImpl::FunctionCall)?;
+                    if convert_comparator_to_priority(comparator_expr.get_expr_type()) < 0 {
+                        return Err(ExprError::InvalidParam {
+                            name: "now",
+                            reason: String::from("now expression must be placed in a comparison"),
+                        }
+                        .into());
+                    }
+                    try_match_expand!(&comparator_expr.inputs()[0], ExprImpl::InputRef)?;
+                    let now_expr =
+                        try_match_expand!(&comparator_expr.inputs()[1], ExprImpl::FunctionCall)?;
+                    match now_expr.get_expr_type() {
+                        Type::Now => {
+                            // Do nothing.
+                        }
+                        Type::Add | Type::Subtract => {
+                            if try_match_expand!(&now_expr.inputs()[0], ExprImpl::FunctionCall)?
+                                .get_expr_type()
+                                != Type::Now
+                            {
+                                return Err(ExprError::InvalidParam {
+                                    name: "now",
+                                    reason: String::from("now() must appear in the left side of the now delta expression directly"),
+                                }
+                                .into());
+                            }
+                        }
+                        _ => {
+                            return Err(ExprError::InvalidParam {
+                                name: "now",
```

Comment:
Should use Optimizer error?

Reply:
> Should use Optimizer error?

I didn't find optimizer error

Reply:
> 😅😇

this is in `to_stream`

Reply:
anyway, just a note, not very serious to be fixed immediately

Reply:
> anyway, just a note, not very serious to be fixed immediately

Do you mean that we can add a new category in `ExprError`, i.e. `ExprError::OptimizerError`?

Reply:
No. IMHO ExprError should only be created in the expr crate. Optimizer/Fronted should have its own error. But currently it's messy, so it's ok not to fix it.

Reply:
> No. IMHO ExprError should only be created in the expr crate. Optimizer/Fronted should have its own error. But currently it's messy, so it's ok not to fix it.

ok let me add a todo